server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java

```diff
@@ -359,6 +360,51 @@ private void initializeInterrupt(ExternalEventModel trigger, InterruptDefModel i
         wfRun.pendingInterrupts.add(pi);
     }
 
+    /**
+     * Attempts to "rescue" the ThreadRun. If not possible, returns a `Status` that can
+     * be thrown to the client explaining why the ThreadRun could not be rescued.
+     * @param skipCurrentNode whether to skip past the current node. If set to `false`, then
+     * we attempt to execute the same Node again; else, we move to the next outgoing edge.
+     * @param ctx is a ProcessorExecutionContext.
+     * @return Optional.empty() if we can successfully rescue the ThreadRun; else, a Status
+     * explaining why not.
+     */
+    public Optional<Status> rescue(boolean skipCurrentNode, ProcessorExecutionContext ctx) {
+        // First, Optional<Status> refers to the grpc status which can be thrown as an error
+        // to the client.
+
+        // Note that any child ThreadRuns that were HALTED with the reason PARENT_HALTED
+        // will be automatically un-halted when the status of this ThreadRun moves from
+        // ERROR to RUNNING
+        if (this.status != LHStatus.ERROR) {
+            throw new IllegalStateException("This is a bug: ThreadRun %s on WfRun %s tried to be rescued from status %s"
+                    .formatted(number, wfRun.getId(), status));
+        }
+
+        try {
+            NodeRunModel currentNR = getCurrentNodeRun();
+            NodeModel toActivate;
+            if (skipCurrentNode) {
+                toActivate = currentNR.evaluateOutgoingEdgesAndMaybeMutateVariables(ctx);
+            } else {
+                toActivate = currentNR.getNode();
+            }
+            setStatus(LHStatus.RUNNING);
+            activateNode(toActivate);
```

Comment:
it seems that now we can have more than 1 NodeRun for an specific Node. is this correct

Reply:
This is correct. But it has been true for a while, since we can have loops in your ThreadSpec which cause the ThreadRun to arrive at a given Node multiple times.

The alternative is to discard the old NodeRun and run it again, but I think that is sloppy since it would be making something mutable that was previously immutable.